canonical_mutation.cc

```diff
@@ -77,7 +78,9 @@ mutation canonical_mutation::to_mutation(schema_ptr s) const {
 
     mutation m(std::move(s), std::move(pk));
 
-    if (version == m.schema()->version()) {
+    // We have to check CDC feature here because until cluster starts supporting
+    // CDC, schema used to exchange info between nodes does not contain CDC column in scylla_tables
+    if (version == m.schema()->version() && service::get_local_storage_service().cluster_supports_cdc()) {
         auto partition_view = mutation_partition_view::from_view(mv.partition());
```

Comment:
canonical_mutation is a primitive, it shouldn't know anything about cluster_supports_cdc(). We should build complicated stuff from primitive stuff, not push special cases into primitives.

I don't understand how that equality test succeeded. If we're sending a schema from a node that has cdc to a node that doesn't have cdc, the schema versions should be different. So I think that check is broken somehow, and you compenated for it by adding the extra check.

Reply:
My suspicious is that a new node (with CDC) sends a canonical mutation to another new node (with CDC). They both have the same schema but before sending CDC column gets removed from the mutation. That's probably why the version of the schema matches. Will check this by testing.

Reply:
Ok, the schema matches, but then what's the problem? It should just work.

We won't pass the cdc column, but nobody should care about it.

Reply:
(we should have debug mode add the schema_ptr to mutation objects and check them on any access, that would remove a whole class of such bugs)

Reply:
(not mutation, but the various mutation_fragments that don't have schemas attached)

Reply:
> Ok, the schema matches, but then what's the problem? It should just work.
> 
> We won't pass the cdc column, but nobody should care about it.

If we try to deserialize the mutation that does not have cdc column with schema that has cdc column, it takes the value of version column as value of cdc column and puts garbage in version column.

Reply:
In which call?

It can't be canonical_mutation::to_mutation(), because that explicitly checks that the schemas match.

Reply:
> My suspicious is that a new node (with CDC) sends a canonical mutation to another new node (with CDC). They both have the same schema but before sending CDC column gets removed from the mutation. That's probably why the version of the schema matches. Will check this by testing.

How confident are you in this?

Reply:
Note that when a column gets removed, the schema version changes:

```c++
static
mutation
redact_columns_for_missing_features(mutation m, schema_features features) {
    if (features.contains(schema_feature::CDC_OPTIONS)) {
        return std::move(m);
    }
    if (m.schema()->cf_name() != SCYLLA_TABLES) {
        return std::move(m);
    }
    slogger.debug("adjusting schema_tables mutation due to possible in-progress cluster upgrade");
    m.upgrade(scylla_tables(features));
    return std::move(m);
}
```

mutation::upgrade() updates the schema of the mutation.

Reply:
> > Ok, the schema matches, but then what's the problem? It should just work.
> > We won't pass the cdc column, but nobody should care about it.
> 
> If we try to deserialize the mutation that does not have cdc column with schema that has cdc column, it takes the value of version column as value of cdc column and puts garbage in version column.

I don't see how this can happen. We use a visitor pattern to deserialize mutations, and thus it is the content of the serialized mutation that drives the deserialization process. How can we visit cells that are not there?

Reply:
> (we should have debug mode add the schema_ptr to mutation objects and check them on any access, that would remove a whole class of such bugs)

There was a patch series posted to do just that: "Catch invalid schema version access on rows", I need to resurrect it.

Reply:
I added some prints to canonical_mutation::to_mutation:
```
mutation canonical_mutation::to_mutation(schema_ptr s) const {
    auto in = ser::as_input_stream(_data);
    auto mv = ser::deserialize(in, boost::type<ser::canonical_mutation_view>());

    auto cf_id = mv.table_id();
    if (s->id() != cf_id) {
        throw std::runtime_error(format("Attempted to deserialize canonical_mutation of table {} with schema of table {} ({}.{})",
                                        cf_id, s->id(), s->ks_name(), s->cf_name()));
    }

    auto version = mv.schema_version();
    auto pk = mv.key();

    mutation m1(s, pk);
    mutation m2(s, pk);

    {
    auto partition_view = mutation_partition_view::from_view(mv.partition());
    mutation_application_stats app_stats;
    m1.partition().apply(*m1.schema(), partition_view, *m1.schema(), app_stats);
    }
    {
    column_mapping cm = mv.mapping();
    converting_mutation_partition_applier v(cm, *m2.schema(), m2.partition());
    auto partition_view = mutation_partition_view::from_view(mv.partition());
    partition_view.accept(cm, v);
    }
    std::cerr << "haaawk " << version << " <-> " << s->version() << std::endl << std::endl << m1 << std::endl << std::endl << m2 << std::endl;
    return version == s->version() ? m1 : m2;
}
```
and I got:
```
haaawk d15e4b38-4294-3cf3-b7e2-5485bd5c9367 <-> d15e4b38-4294-3cf3-b7e2-5485bd5c9367

{system_schema.scylla_tables {key: pk{00026b73}, token:3274692326281147944} {mutation_partition:
 static: cont=1 {row: },
 clustered: {
    {rows_entry: cont=true dummy=false {position: clustered,ckp{00026366},0} {deletable_row: {row:
      {0, { atomic_cell{7aa08101217911eab225000000000000;ts=1576661622032685;expiry=-1,ttl=0} }}}}}}}}

{system_schema.scylla_tables {key: pk{00026b73}, token:3274692326281147944} {mutation_partition:
 static: cont=1 {row: },
 clustered: {
    {rows_entry: cont=true dummy=false {position: clustered,ckp{00026366},0} {deletable_row: {row:
      {1, { atomic_cell{7aa08101217911eab225000000000000;ts=1576661622032685;expiry=-1,ttl=0} }}}}}}}}
```

The cannonical_mutation that the method is called on is a mutation without CDC column received from another node.
I performed both normal deserialization and schema conversion to compare the results.
As you can see, schema version of canonical mutation (that does not contain CDC column) and schema version on the node (that contains CDC column) are the same so the method "thinks" that it should perform deserialization without schema conversion.
Is |s|, column 0 is CDC column and 1 is version column.
cannonical_mutation contains only data for version column.
You can see that deserialization without schema conversion places data from cannonical mutation into CDC column.
Deserialization with schema conversion places data from cannonical mutation into version column, correctly.

So the problem is that when we exchange schema with other nodes and remove a column, the version of the schema stays the same.

Any thoughts, @avikivity ?

Reply:
That's why comparing not only schema version but also CDC feature solved the problem.
I understand @avikivity why you didn't like the approach so would you rather accept the previous solution that was removing CDC column from |s| before passing it to cannonical_mutation::to_mutation in migration_manager?

Reply:
That's the bug for sure, now to understand how it happened.

Reply:
It's this line:

```
            .with_version(generate_schema_version(id));
```

Reply:
So I propose to make the version input "1" without cdc, and "2" after cdc (and make generate_schema_version accept version_sequence_number as a parameter instead of reading it from the global version_sequence_number)

Reply:
@tgrabiec?

Reply:
> That's why comparing not only schema version but also CDC feature solved the problem.
> I understand @avikivity why you didn't like the approach so would you rather accept the previous solution that was removing CDC column from |s| before passing it to cannonical_mutation::to_mutation in migration_manager?

Can you refer me to that? I have no idea what |s| is.

Reply:
> > That's why comparing not only schema version but also CDC feature solved the problem.
> > I understand @avikivity why you didn't like the approach so would you rather accept the previous solution that was removing CDC column from |s| before passing it to cannonical_mutation::to_mutation in migration_manager?
> 
> Can you refer me to that? I have no idea what |s| is.

|s| is schema_ptr taken by canonical_mutation::to_mutation as parameter. I think fixing generate_schema_version is a better solution though.

Reply:
There's this comment on generate_schema_version:
```
// Increase whenever changing schema of any system table.
// FIXME: Make automatic by calculating from schema structure.
static const uint16_t version_sequence_number = 1;

table_schema_version generate_schema_version(utils::UUID table_id) {
    md5_hasher h;
    feed_hash(h, table_id);
    feed_hash(h, version_sequence_number);
    return utils::UUID_gen::get_name_UUID(h.finalize());
}
```
Maybe I just need to increase version_sequence_number?

Reply:
It seems that increasing version_sequence_number helped but I need to run more tests to be sure. Tests that don't fail don't prove anything ;).

Reply:
When I think about it, I can't explain why would it help though. Both nodes will be using increased version_sequence_number so the schema version should be the same again and lead to the same problem.

Reply:
Yes, it needs to be dependent on schema_features.

Reply:
also, I wouldn't want to change the sequence number for all tables needlessly.

Reply:
Maybe it worked because the schema registry happened to return the right schema for the version this time around (we register two schemas with the same version, so it is implementation-defined which schema we get when we look up by version)

Reply:
Ok. So maybe we could take an additional parameter to generate_schema_version that is an offset above version_sequence_number?

Reply:
Makes sense. version_sequence_number must be frozen for this, but it effectively is.